grpc/execution/server.go

```diff
@@ -123,11 +129,226 @@ func (s *ExecutionServiceServer) FinalizeBlock(ctx context.Context, req *executi
+	}
+
+	// This set of ordered TXs on the TxPool is has been configured to be used by
+	// the Miner when building a payload.
+	s.eth.TxPool().SetAstriaOrdered(req.Transactions)
+
+	// Build a payload to add to the chain
+	payloadAttributes := &miner.BuildPayloadArgs{
+		Parent:       prevHeadHash,
+		Timestamp:    uint64(req.GetTimestamp().GetSeconds()),
+		Random:       common.Hash{},
+		FeeRecipient: common.Address{},
+	}
+	payload, err := s.eth.Miner().BuildPayload(payloadAttributes)
+	if err != nil {
+		log.Error("failed to build payload", "err", err)
+		return nil, status.Error(codes.InvalidArgument, "Could not build block with provided txs")
+	}
+
+	// call blockchain.InsertChain to actually execute and write the blocks to
+	// state
+	block, err := engine.ExecutableDataToBlock(*payload.Resolve().ExecutionPayload)
+	if err != nil {
+		return nil, err
+	}
+	blocks := types.Blocks{
+		block,
+	}
+	n, err := s.bc.InsertChain(blocks)
+	if err != nil {
+		return nil, status.Error(codes.Internal, "failed to insert block to chain")
+	}
+	if n != 1 {
+		log.Error("block was inserted at height ", n, " instead of head")
+		return nil, status.Error(codes.Internal, "Failed to insert block to chain")
+	}
+
+	// remove txs from original mempool
+	for _, tx := range block.Transactions() {
+		s.eth.TxPool().RemoveTx(tx.Hash())
+	}
+
+	res := &executionv1a2.Block{
+		Number: uint32(block.NumberU64()),
+		Hash:   block.Hash().Bytes(),
+		Timestamp: &timestamppb.Timestamp{
+			Seconds: int64(block.Time()),
+		},
+	}
+	return res, nil
+}
+
+// GetCommitmentState fetches the current CommitmentState of the chain.
+func (s *ExecutionServiceServerV1Alpha2) GetCommitmentState(ctx context.Context, req *executionv1a2.GetCommitmentStateRequest) (*executionv1a2.CommitmentState, error) {
+	softBlock, err := ethHeaderToExecutionBlock(s.bc.CurrentSafeBlock())
+	firmBlock, err := ethHeaderToExecutionBlock(s.bc.CurrentFinalBlock())
+
+	if err != nil {
+		return nil, err
+	}
+
+	res := &executionv1a2.CommitmentState{
+		Soft: softBlock,
+		Firm: firmBlock,
+	}
+
+	return res, nil
+}
+
+// UpdateCommitmentState replaces the whole CommitmentState with a new
+// CommitmentState.
+func (s *ExecutionServiceServerV1Alpha2) UpdateCommitmentState(ctx context.Context, req *executionv1a2.UpdateCommitmentStateRequest) (*executionv1a2.CommitmentState, error) {
+	softEthHash := common.BytesToHash(req.CommitmentState.Soft.Hash)
+	firmEthHash := common.BytesToHash(req.CommitmentState.Firm.Hash)
+
+	// Validate that the firm and soft blocks exist before going further
+	softBlock := s.bc.GetBlockByHash(softEthHash)
+	if softBlock == nil {
+		return nil, status.Error(codes.InvalidArgument, "Soft block specified does not exist")
+	}
+	firmBlock := s.bc.GetBlockByHash(firmEthHash)
+	if firmBlock == nil {
+		return nil, status.Error(codes.InvalidArgument, "Firm block specified does not exist")
+	}
+
+	currentHead := s.bc.CurrentBlock().Hash()
```

Comment:
```suggestion
	currentHead := s.bc.CurrentBlock()
```

then at like 299 you don't need to fetch the block again

Reply:
Quite counter intuitively, `CurrentBlock` returns a header, so I still need to grab the whole block.